### ISSUE TITLE
Write script to compare set.mm and iset.mm

### ISSUE THREAD

**Issue by jkingdon** [metamath/set.mm#2891]
We have a convention which is stated as

> Theorems which are the same as in set.mm should be named the same (that is, where the statement of the theorem is the same; the proof can differ without a new name being called for). Theorems which are different should be named differently (we do have a small number of intentional exceptions to this rule but on the whole it serves us well).

in https://us.metamath.org/ileuni/conventions.html

The proposed script would check that this is case and print the names of theorems which do not fit this rule.

Details:
* The script could contain a static list of the "small number of intentional exceptions"
* "statement of the theorem is the same" can be a character-by-character comparison of `$e` hypotheses and the statement (not proof) from `$p`. That is, it is OK (perhaps even desirable) to consider renamed variables as differences.
* The easiest thing to do about distinct variable constraints would be to ignore them. In this scenario we'd need a second exception list for cases like https://us.metamath.org/mpeuni/nfral.html and https://us.metamath.org/ileuni/nfralxy.html , which have different names because they have different distinct variable constraints (the number of theorems like this is probably quite small). A more rigorous approach would be to compare distinct variable constraints from non-dummy variables (including dummy variables would not work well, because they often differ based on the proof being different, and the proof being different does not call for a different name).
* Another possibility (although this is perhaps better separated from the other topics, see discussion in the comments) we could add to the script (and, I suppose, to conventions) is that if a theorem is in main in either iset.mm or set.mm, it should be in main (as opposed to a mathbox) in the other. The idea being that we'd move theorems from mathboxes to main when we find this isn't the case.

**Comment by benjub** [comment]
Yes, that would be a nice script to have.

As for your last point (moving from a mathbox to Main any theorem which is in the main part of one of iset.mm or set.mm) I fully agree.

As for what "same statement" means, I would say that it is "up to variable renaming". This makes coding a bit more difficult. One can of course put an ordering on the set of all formulas which are the same up to variable renaming, so we have a "normal form" for comparisons. And one can first ignore DVs, as you propose.

A comparision between theorems in iset.mm and set.mm would be feasible, I guess: one can put an ordering on the space of all formulas (it could be as simple as the lexicographic order, totally blind to formula meaning), so then time complexity of the comparison is O(m+n) where m, n are the number of statements in each database, after they have been sorted, so O(m.log(m)  + m log(m)).

(I hope that tokens have the same name in set and iset?  e.g., conjunction is `/\` in both, etc.)

**Comment by jkingdon** [comment]
> As for what "same statement" means, I would say that it is "up to variable renaming"

If there are statements which are the same in iset.mm and set.mm but have different variable names, I would prefer to adopt the same names both places (usually by having iset.mm adopt the set.mm names). This is a minor point in the grand scheme of things, but it might be a case where the thing which is easier to code is also the more desirable behavior.

> A comparision between theorems in iset.mm and set.mm would be feasible, I guess: one can put an ordering

I think that algorithm could work or it probably would be possible to build up an index. That is, store all the theorems from one database in a python dictionary or similar data structure, and then scan through the theorems from the other database and look them up in that dictionary which would be fast (this is perhaps another place where a choice to do "up to renaming" complicates things - although the formation of an index key from the statement could canonicalize the names if need be).

> (I hope that tokens have the same name in set and iset? e.g., conjunction is /\ in both, etc.)

I'm not thinking of any exceptions. Nor did I see any browsing through the htmldef lines. Perhaps the closest is the additional argument to `seq` (not a new token, but a new syntax using existing tokens) but we are in the process of (slowly) removing that one.

**Comment by benjub** [comment]
Yes, I agree: if statements differ only by variable renaming, then the script should signal them, and then one should harmonize both the labels and the variable names.

**Comment by jkingdon** [comment]
> As for your last point (moving from a mathbox to Main any theorem which is in the main part of one of iset.mm or set.mm) I fully agree.

That's actually the one I'm most uncertain about. Theorems have various reasons for moving between mathboxes and main and although I guess maybe we could always do it together in both set.mm and iset.mm, I'm a little bit worried about making tricky situations more tricky (not that I have a currently active example in mind).

**Comment by jkingdon** [comment]
One piece of prior art which I just noticed but which still runs and produces interesting output is https://github.com/metamath/set.mm/blob/develop/scripts/iset-match.py .

I don't know if it is productive to past the entire output here, but it does still run:

```
$ python3 scripts/iset-match.py  | wc -l
207
$
```
